host/sgx/sgxquote.c

```diff
@@ -125,18 +163,214 @@ static bool _load_sgx_dcap_ql(void)
+            local_key_id, att_key_id_num);
+        if (status != SGX_SUCCESS)
+            OE_RAISE_MSG(
+                OE_PLATFORM_ERROR,
+                "_load_quote_ex_library_once() "
+                "sgx_get_supported_att_key_ids() status=%d\n",
+                status);
+
+        for (uint32_t i = 0; i < att_key_id_num; i++)
+        {
+            sgx_att_key_id_ext_t* key = local_key_id + i;
+            const oe_uuid_t* uuid = NULL;
+
+            OE_TRACE_INFO("algorithm_id=%d", key->base.algorithm_id);
+
+            switch (key->base.algorithm_id)
+            {
+                case SGX_QL_ALG_EPID_UNLINKABLE:
+                    uuid = &_epid_unlinkable_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                case SGX_QL_ALG_EPID_LINKABLE:
+                    uuid = &_epid_linkable_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                case SGX_QL_ALG_ECDSA_P256:
+                    uuid = &_ecdsa_p256_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                case SGX_QL_ALG_ECDSA_P384:
+                    uuid = &_ecdsa_p384_uuid;
+                    local_mapped[i] = true;
+                    mapped_key_id_count++;
+                    break;
+                default:
+                    uuid = &_unknown_uuid;
+                    local_mapped[i] = false;
+                    OE_TRACE_ERROR(
+                        "algorithm_id=%d maps to no uuid",
+                        key->base.algorithm_id);
+                    break;
+            }
+            memcpy(local_uuid + i, uuid, sizeof(oe_uuid_t));
+        }
+
+        _quote_ex_library.key_id_count = att_key_id_num;
+        _quote_ex_library.mapped_key_id_count = mapped_key_id_count;
+        _quote_ex_library.mapped = local_mapped;
+        _quote_ex_library.uuid = local_uuid;
+        _quote_ex_library.sgx_key_id = local_key_id;
+        local_mapped = NULL;
+        local_uuid = NULL;
+        local_key_id = NULL;
+
+        OE_TRACE_INFO(
+            "key_id_count=%lu mapped=%lu\n",
+            att_key_id_num,
+            mapped_key_id_count);
+
+        result = OE_OK;
+    }
+
+done:
+    if (local_mapped)
+    {
+        oe_free(local_mapped);
+        local_mapped = NULL;
+    }
+    if (local_uuid)
+    {
+        oe_free(local_uuid);
+        local_uuid = NULL;
+    }
+    if (local_key_id)
+    {
+        oe_free(local_key_id);
+        local_key_id = NULL;
+    }
+    if (_quote_ex_library.load_result == OE_OK)
+        _quote_ex_library.load_result = result;
+
+    OE_TRACE_INFO(
+        "_load_quote_ex_library_once() result=%s\n",
+        oe_result_str(_quote_ex_library.load_result));
+
+    return;
+}
+
+// For choosing between the DCAP library and the quote-ex library for quote
```

Comment:
(Quoting a dislocated comment for context)
> @shnwc I don't think that speaks to @yentsanglee's suggestion, which isn't about why there are two methods of getting the target info in this function. I think he's pointing out that it's not obvious from reading the code why _sgx_qe_get_target_info() is only called if oe_sgx_initialize_quote_ex_library() fails entirely. i.e. it looks like an app expecting to talk to the DCAP in-proc QE would get short-circuited and never called if the quote_ex library is loadable.
>
> It would be useful to explain as a code comment here (and elsewhere where this pattern is repeated, like in oe_sgx_qe_get_quote_size()) why this is either acceptable, or not the case (e.g. we expect that oe_sgx_initialize_quote_ex_library() is the function that checks the whether or not to use an internal provider and will fail so that this code then falls through to the _sgx_qe_get_target_info() call?).

Thanks for adding the comment, but it doesn't explain _why_ this choice exists to a new maintainer reading this code, or why this order of fallback is chosen. For example, why not try `_load_sgx_dcap_ql()` first and then fallback to `__use_quote_ex_library()` if that fails? They're both effectively just load library checks. Given the current logic, if I want to use the existing DCAP library, I cannot install the quote_ex library on the same device even though there may be two different host apps that may choose to use libraries; the host app doesn't seem to get a choice, only the runtime environment matters.

Reply:
@CodeMonkeyLeet this PR implements the design proposal on [OE SDK integration with the Intel SGX SDK quote-ex library](../blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md), with update of the design doc in PR #3329.

If a reference of the design doc is included in the proposed `3rdparty/sgxsdk/README.md`, will this be sufficient for clarification? Or do you see some other files where this design doc should be referenced?

Reply:
The original design proposed a different linking and selection mechanism, which has since changed. Option 1 was only briefly documented and not deeply considered in the original design because it was not the recommended option and hence not the focus of that design.

Now that this PR is choosing to implement that option instead, I would like to ensure that I understand the developer use case. The question that I am asking here needs to be answered as part of the proposed update to that design #3329.

Reply:
Will add a reference to the design document as part of this comment.

Reply:
> The original design proposed a different linking and selection mechanism, which has since changed. Option 1 was only briefly documented and not deeply considered in the original design because it was not the recommended option and hence not the focus of that design.
> 
> Now that this PR is choosing to implement that option instead, I would like to ensure that I understand the developer use case. The question that I am asking here needs to be answered as part of the proposed update to that design #3329.

The implementation moved from the original option 4 to the latest option 1 due to the merge of another PR #3273. The priority between the DCAP and the quote-ex libraries stays the same between the two options.